### PR TITLE
Collect cgroup v2 memory stats

### PR DESCRIPTION
**Pull request by EricYangIBM**
Update `retrieveLinuxCgroupMemoryStats` to collect memory metrics from
cgroup v2 files.
The cgroup memory files are changed from cgroup v1 to v2 as follows:
`memory.limit_in_bytes` -> `memory.max`
`memory.usage_in_bytes` -> `memory.current`
`memory.memsw.limit_in_bytes` -> `memory.swap.max`
`memory.memsw.usage_in_bytes` -> `memory.swap.current`
`memory.stat` no change
In the `memory.stat` file, the metric `cache` -> `file`

Issue: https://github.com/eclipse/omr/issues/1281
Signed-off-by: Eric Yang <eric.yang@ibm.com>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
@babsingh : please review

**Comment by babsingh** [comment]
Currently, this PR also includes changes from https://github.com/eclipse/omr/pull/6432. Once https://github.com/eclipse/omr/pull/6432 is merged, this PR will need to be rebased. Then, it will be ready for a review.

**Review by babsingh — CHANGES_REQUESTED** [review]
Does `PortSysinfoTest, sysinfo_testMemoryInfo`, which invokes `omrsysinfo_get_memory_info`, work correctly on a cgroup v2 machine? Can you post its output?

**Comment by EricYangIBM** [comment]
I think the `porttest`s are not being run:
```
test 6
      Start  6: porttest

6: Test command: /root/docker_hostdir/openj9-openjdk-jdk8/omr/build/fvtest/porttest/omrporttest "--gtest_output=xml:/root/docker_hostdir/openj9-openjdk-jdk8/omr/build/fvtest/porttest/omrporttest-results.xml" "--gtest_filter="
6: Test timeout computed to be: 10000000
6: Note: Google Test filter = 
6: [==========] Running 0 tests from 0 test cases.
6: [==========] 0 tests from 0 test cases ran. (0 ms total)
6: [  PASSED  ] 0 tests.
6: [  ALL TESTS PASSED  ] 
 6/17 Test  #6: porttest .........................   Passed    0.01 sec
```

**Comment by babsingh** [comment]
> re https://github.com/eclipse/omr/pull/6465#issuecomment-1104024725: I think the porttests are not being run:

@EricYangIBM How are you running the test (cmd + environment)?

**Comment by EricYangIBM** [comment]
I followed the instructions in https://github.com/eclipse/omr/blob/master/doc/BuildingWithCMake.md#running-tests (running on host vm)

**Comment by babsingh** [comment]
I rebased my OMR repo, and I was able to reproduce the error, which is mentioned in https://github.com/eclipse/omr/pull/6465#issuecomment-1104024725. The introduction of `"--gtest_filter="` at the end of the `Test command` causes 0 `port tests` to run. It is apparently affecting all PR builds. None of the PR builds are correctly running `porttest`.

### With the error
```
/root/omr/build/fvtest/porttest/omrporttest "--gtest_output=xml:/root/omr/build/fvtest/porttest/omrporttest-results.xml" "--gtest_filter="
Note: Google Test filter =
[==========] Running 0 tests from 0 test cases.
[==========] 0 tests from 0 test cases ran. (0 ms total)
[  PASSED  ] 0 tests.
[  ALL TESTS PASSED  ]
```

### Without the error
```
/root/omr/build/fvtest/porttest/omrporttest "--gtest_output=xml:/root/omr/build/fvtest/porttest/omrporttest-results.xml"
[==========] Running 237 tests from 20 test cases.
[----------] 1 test from PortInitializationTest
[----------] 1 test from PortInitializationTest (0 ms total)

[----------] 2 tests from PortFileTest
[----------] 2 tests from PortFileTest (0 ms total)

[----------] 8 tests from PortHeapTest
[----------] 8 tests from PortHeapTest (1099 ms total)

[----------] 8 tests from PortTest
...
```

**Comment by babsingh** [comment]
re https://github.com/eclipse/omr/pull/6465#issuecomment-1104252946:

Before we merge this PR, there needs to be verification that the new code works correctly. If there is an easier (alternate) way to verify, we can update the test in a later PR. Otherwise, we should either update the existing test or add a new test, to verify cgroup code, as part of this PR.

File: port/unix/omrsysinfo.c [hunk 1]
```diff
@@ -3074,31 +3091,49 @@ retrieveLinuxCgroupMemoryStats(struct OMRPortLibrary *portLibrary, struct OMRCgr
...
 	cgroupMemInfo->cached = OMRPORT_MEMINFO_NOT_AVAILABLE;
 
-	rc = readCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_MEMORY, CGROUP_MEMORY_LIMIT_IN_BYTES_FILE, numItemsToRead, "%lu", &cgroupMemInfo->memoryLimit);
+	if (OMR_ARE_ANY_BITS_SET(PPG_sysinfoControlFlags, OMRPORT_SYSINFO_CGROUP_V1_AVAILABLE)) {
+		memLimitFile = CGROUP_MEMORY_LIMIT_IN_BYTES_FILE;
+		memUsageFile = CGROUP_MEMORY_USAGE_IN_BYTES_FILE;
+		swpLimitFile = CGROUP_MEMORY_SWAP_LIMIT_IN_BYTES_FILE;
+		swpUsageFile = CGROUP_MEMORY_SWAP_USAGE_IN_BYTES_FILE;
+		metricName = CGROUP_MEMORY_STAT_CACHE_METRIC;
+		metricSize = CGROUP_MEMORY_STAT_CACHE_METRIC_SZ;
```
Comment (babsingh) [inline port/unix/omrsysinfo.c:3100]: `metricName` and `metricSize` are generic. They should be given a meaningful name. `cachedMetricName` and `cachedMetricNameSize`.

**Comment by EricYangIBM** [comment]
I can verify that the struct values returned by `retrieveLinuxCgroupMemoryStats` are the same as what's in the cgroup memory files
```
memoryLimit 4294967296
memoryUsage 12156928
memoryAndSwapLimit 1073741824
memoryAndSwapUsage 0
cached 8650752
```
```
root@deb739a1b00c:~/hostdir/temp/omr/build# cat /sys/fs/cgroup/memory.max
4294967296
root@deb739a1b00c:~/hostdir/temp/omr/build# cat /sys/fs/cgroup/memory.current 
12435456
root@deb739a1b00c:~/hostdir/temp/omr/build# cat /sys/fs/cgroup/memory.swap.max 
1073741824
root@deb739a1b00c:~/hostdir/temp/omr/build# cat /sys/fs/cgroup/memory.swap.current 
0
root@deb739a1b00c:~/hostdir/temp/omr/build# cat /sys/fs/cgroup/memory.stat | grep file
file 8785920
```
This is with docker command `docker run -v /root/docker_hostdir:/root/hostdir -it -m 4GB --memory-swap=5GB  --rm openj9`

**Comment by babsingh** [comment]
These changes LGTM. Waiting for https://github.com/eclipse/omr/pull/6487 to be merged so that `porttest` is correctly run in the PR builds.

**Review by babsingh — CHANGES_REQUESTED** [review]
Also describe translations from cgroup v1 to cgroup v2 in the commit message:
- "memory.limit_in_bytes" translates to "memory.max".
- "memory.usage_in_bytes" translates to "memory.current".
- "memory.memsw.limit_in_bytes" translates to "memory.swap.max".
- "memory.memsw.usage_in_bytes" translates to "memory.swap.current".
- "memory.stat" stays the same.
- In "memory.stat", metric "cache" translates to "file".

File: port/unix/omrsysinfo.c [hunk 2]
```diff
@@ -3074,31 +3091,49 @@ retrieveLinuxCgroupMemoryStats(struct OMRPortLibrary *portLibrary, struct OMRCgr
...
+	} else {
+		Trc_PRT_Assert_ShouldNeverHappen();
+	}
+
+	rc = readCgroupMemoryFileIntOrMax(portLibrary, memLimitFile, &cgroupMemInfo->memoryLimit);
 	if (0 != rc) {
 		goto _exit;
 	}
-	rc = readCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_MEMORY, CGROUP_MEMORY_USAGE_IN_BYTES_FILE, numItemsToRead, "%lu", &cgroupMemInfo->memoryUsage);
+	rc = readCgroupMemoryFileIntOrMax(portLibrary, memUsageFile, &cgroupMemInfo->memoryUsage);//
```
[the remark this answers is not in the document]
Comment (EricYangIBM) [reply port/unix/omrsysinfo.c]: I left that as a reminder to revert the change since the .current files can't have "max" in them but forgot

**Comment by babsingh** [comment]
Ref: https://github.com/containers/crun/blob/main/crun.1.md#memory-controller

> the swap limit on cgroup v1 includes the memory usage too

We will have to account for the diff in `cgroupMemInfo->memoryAndSwapLimit`. So, the values are identical in cgroup v1 and v2.

**Comment by EricYangIBM** [comment]
Got v2 working correctly
```
root@aeeb43efc1aa:~/hostdir/openj9-openjdk-jdk8# build/linux-x86_64-normal-server-release/images/j2sdk-image/bin/java -version
memoryLimit 4294967296
memoryUsage 24821760
memoryAndSwapLimit 5368709120
memoryAndSwapUsage 24821760
```

**Comment by babsingh** [comment]
> Zos git clone failure on this PR and https://github.com/eclipse/omr/pull/6479. Is this a known issue?

Yes, infra issue: `Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to ZISVJD10`. These changes only impact Linux builds. So, the zOS is not critical for these changes.

**Review by babsingh — APPROVED** [review]
@0xdaryl On my end, this PR is good to merge. Requesting your feedback.

**Comment by babsingh** [comment]
jenkins build zos

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**